File: prototype/rails/src/app/domain/shop/checkout_purchaser.rb
require_relative "../auth/email_address"
require_relative "../orders/purchaser"

module Domain
  module Shop
    module CheckoutPurchaser
      module_function

      # A signed-in customer buys under the address on their account, so a
      # submitted field cannot move an order onto someone else's identity. A
      # guest buys under the address they typed and verifies it afterwards.
      def for_checkout(id:, account_email:, account_verified_at:, submitted_email:)
        return Orders::Purchaser.new(id: id, email: account_email, email_verified_at: account_verified_at) if
          account_verified_at

        Orders::Purchaser.new(
          id: id, email: Auth::EmailAddress.normalize(submitted_email), email_verified_at: nil
        )
      end
    end
  end
end

File: prototype/node/src/app/core/shop/checkout-purchaser.ts
import { normalizeEmail } from '../auth/email-address.ts'
import type { Purchaser } from '../orders/purchaser.ts'
import type { CustomerId } from '../ids/entity-ids.ts'

export type CheckoutIdentity = {
  customerId: CustomerId
  accountEmail: string | null
  isAccountVerified: boolean
  submittedEmail: string
}

/**
 * A signed-in customer buys under the address on their account, so a submitted
 * field cannot move an order onto someone else's identity. A guest buys under
 * the address they typed and verifies it afterwards.
 */
export function purchaserForCheckout(identity: CheckoutIdentity): Purchaser {
  if (identity.isAccountVerified && identity.accountEmail !== null) {
    return { id: identity.customerId, email: identity.accountEmail, isEmailVerified: true }
  }

  return {
    id: identity.customerId,
    email: normalizeEmail(identity.submittedEmail),
    isEmailVerified: false,
  }
}
